### PR TITLE
Detect TP element orientation, add support for TP flipping

### LINKED ISSUE #403: 3d element orientation for hexes
I've attached a gmsh mesh and .geo file used to generate it. This mesh has a node ordering that is inconsistent with the lexicographic vertex reshuffle for hexahedral elements used in meshmode. 

An example run on the mesh in question:

image
image

if I modify the vertex ordering by swapping 1 with 4 and 7 with 2 I can produce a working mesh

image

### PR DESCRIPTION
**Pull request by inducer**
Closes #403.

- [x] Add to "neg orient" error message how to fix

### CODE REVIEW

File: meshmode/mesh/__init__.py
Comment (inducer) [inline meshmode/mesh/__init__.py:1]: @alexfikl Since you last touched this error reporting code, could you take a quick look whether you like this?

Comment (alexfikl) [reply meshmode/mesh/__init__.py:1]: :+1: Definitely! I wanted to remove that `test_volume_mesh_element_orientations` function too.

File: meshmode/mesh/processing.py [hunk 1]
```diff
@@ -619,43 +642,31 @@ def find_volume_mesh_element_orientations(
...
     return result
 
-
-def test_volume_mesh_element_orientations(mesh: Mesh) -> bool:
-    area_elements = find_volume_mesh_element_orientations(
-            mesh, tolerate_unimplemented_checks=True)
-    valid = ~np.isnan(area_elements)
-
-    return bool(np.all(area_elements[valid] > 0))
-
```
[the remark this answers is not in the document]
Comment (alexfikl) [reply meshmode/mesh/processing.py:648]: Looks good!

File: meshmode/mesh/__init__.py
Comment (alexfikl) [reply meshmode/mesh/__init__.py:1]: Is it worth adding a `perform_element_flips=False` flag to `make_mesh` to do the flips automagically if possible?

Comment (inducer) [reply meshmode/mesh/__init__.py:1]: I agree, that's more usable. I'll do that.

Comment (inducer) [reply meshmode/mesh/__init__.py:1]: Done!

File: meshmode/interop/firedrake/connection.py [hunk 2]
```diff
@@ -726,7 +726,7 @@ def build_connection_from_firedrake(actx, fdrake_fspace, grp_factory=None,
     # a numpy array
 
     # Get the reordering fd->mm.
-    flip_mat = get_simplex_element_flip_matrix(ufl_elt.degree(),
+    flip_mat, _perm = get_simplex_element_flip_matrix(ufl_elt.degree(),
                                                fd_unit_nodes)
```
Comment (alexfikl) [inline meshmode/interop/firedrake/connection.py:730]:
```suggestion
    flip_mat, _perm = get_simplex_element_flip_matrix(ufl_elt.degree(),
                                                      fd_unit_nodes)
```
? The one below is also not aligned anymore.

File: meshmode/interop/firedrake/mesh.py [hunk 3]
```diff
@@ -967,7 +967,7 @@ def export_mesh_to_firedrake(mesh, group_nr=None, comm=None):
                        " in firedrake nodal order"):
         from meshmode.mesh.processing import get_simplex_element_flip_matrix
         for perm, cells in perm2cells.items():
-            flip_mat = get_simplex_element_flip_matrix(group.order,
+            flip_mat, _perm = get_simplex_element_flip_matrix(group.order,
                                                        fd_unit_nodes,
                                                        perm)
```
Comment (alexfikl) [inline meshmode/interop/firedrake/mesh.py:972]:
```suggestion
            flip_mat, _perm = get_simplex_element_flip_matrix(group.order,
                                                              fd_unit_nodes,
                                                              perm)
```
?

File: meshmode/mesh/processing.py [hunk 4]
```diff
@@ -1538,4 +1596,36 @@ def make_mesh_grid(
...
+        return vi
+
+    used_vertices = np.unique(np.sort(np.concatenate([
+        not_none(grp.vertex_indices).reshape(-1)
+        for grp in mesh.groups
+    ])))
+
+    used_flags: np.ndarray = np.zeros(mesh.nvertices, dtype=np.bool_)
+    used_flags[used_vertices] = 1
+    new_vertex_indices = np.cumsum(used_flags.astype(mesh.vertex_id_dtype)) - 1
```
Comment (alexfikl) [inline meshmode/mesh/processing.py:1618]:
```suggestion
    new_vertex_indices = np.cumsum(used_flags, dtype=mesh.vertex_id_dtype) - 1
```
? I think that's the same, not sure if it makes any difference..

**Review by alexfikl — COMMENTED** [review]
Left a few nitpicks since I was here :grin:

File: meshmode/mesh/processing.py (hunk 4 above)
Comment (inducer) [reply meshmode/mesh/processing.py]: It actually isn't the same; it seems that's the only way to influence the `cumsum` output dtype.

**Pull request merged**